### PR TITLE
Implement blocking disposable email domains

### PR DESCRIPTION
**Pull request by FawazFarid**
This PR introduces functionality to block registration and usage of disposable email domains within Mastodon. Disposable email domains are often used by bots, spammers or for temporary purposes, and blocking them helps maintain the quality of user data and reduces the risk of spam and abuse.

### Testing:
Tests have been added to cover the new functionality.

### Screenshots
Screenshot 2024-04-17 at 5 13 09 PM

### Impact:

- Users attempting to register with disposable email domains will be prevented from doing so.
- Existing users attempting to update their emails with such domains will be prevented from doing so.

### Some things to think about
We might need to have a setting for admins to opt in/out of this behavior. We'll turn the feature off by default (if we ship it) but let the admins decide whether to turn it on or not.

### CODE REVIEW

**Comment by ThisIsMissEm** [comment]
Could this be better solved by just adding an import capability to the "Blocked e-mail domains" feature? `/admin/email_domain_blocks`

File: lib/tasks/disposable_email_domains.rake [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace :disposable_email_domains do
+  desc 'Download latest list of disposable email domains'
+  task download: :environment do
+    data = HTTP.get('https://disposable.github.io/disposable-email-domains/domains.json').to_s
+
+    dir = Rails.root.join('data')
+    FileUtils.mkdir_p(dir)
+
+    File.write("#{dir}/disposable_email_domains.txt", data, mode: 'w')
```
Comment (ThisIsMissEm) [inline lib/tasks/disposable_email_domains.rake:11]: I'd encourage keeping the file extension the same as the origin, so:
```suggestion
    File.write("#{dir}/disposable_email_domains.json", data, mode: 'w')
```

**Comment by shleeable** [comment]
This makes a lot of sense, but I wouldn't block them... just mark the account as review required

**Comment by ocdtrekkie** [comment]
New disposable domains are rotated pretty regularly by the popular providers, shipping a static list of these (especially at Mastodon's current release cadence) would be pretty useless, wouldn't it?

**Comment by shleeable** [comment]
This is also similar to #29243 - I've tried to add a variable to pull in known risky IPs such as Tor exit relays... 

Didn't seem to get much traction.

**Comment by mjankowski** [comment]
Closing due to lack of activity and rebase needed -- and an assumption that any viable version of this would need to have some updating/downloading capability built-in (as opposed to static).

If you want to nudge forward, can you rebase, re-open, confirm still relevant on latest main, address any feedback, etc...

**Pull request closed without merging**